distributed/scheduler.py

```diff
@@ -5532,14 +5531,72 @@ async def proxy(self, comm=None, msg=None, worker=None, serializers=None):
         )
         return d[worker]
 
-    async def _delete_worker_data(self, worker_address, keys):
+    async def _gather_on_worker(
+        self, worker_address: str, who_has: "dict[Hashable, list[str]]"
+    ) -> set:
+        """Peer-to-peer copy of keys from multiple workers to a single worker
+
+        Parameters
+        ----------
+        worker_address: str
+            Recipient worker address to copy keys to
+        who_has: dict[Hashable, list[str]]
+            {key: [sender address, sender address, ...], key: ...}
+
+        Returns
+        -------
+        returns:
+            set of keys that failed to be copied
+        """
+        result = await retry_operation(
+            self.rpc(addr=worker_address).gather, who_has=who_has
+        )
```

Comment:
What happens if the worker disconnects?

Reply:
I'm working to bake into self.rpc graceful handling for it (in scope for this PR but not yet in - the unit tests for it are already there though).

Reply:
it's now in